Added documentation

naved001 opened the pull request:
What have I added here?

* Installation instructions (copied from the MOC wiki)
* Developer guidelines (based on HIL documentations)

chemistry-sourabh commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,279 @@
+# BMI
+
+The Bare Metal Imaging (BMI) is a core component of the Massachusetts Open Cloud and a image management system(ims) that  
+* provisions numerous nodes as quickly as possible while preserving support for multitenancy using Hardware as a Service (HaaS) and  
```
Change it to HIL. Everywhere

naved001 replied:
Done

chemistry-sourabh commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,279 @@
+***
+## Running
+
+### Creating Tunnel (Only for internal purposes)
```
Remove this from here. Not supposed to be in public repo

apoorvemohan replied:
@chemistry-sourabh what do you mean by "this"?

apoorvemohan replied:
Section "#### iSCSI Server" needs to be edited with correct way to setup TGT

chemistry-sourabh replied:
Creating the tunnel section

naved001 replied:
Done

chemistry-sourabh commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,279 @@
+
+For Northeastern and Engage1 just the regular tunnel.
+
+### Configuration
```
Will need to update this entire Section after PR merge

naved001 replied:
yes, will do that in the next commit.

chemistry-sourabh commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,279 @@
+## Installation
+
+This document describes how to setup BMI
+
```
I advise adding @sirushtim's script as the first way to install for developers. Then the manual installation can continue

naved001 replied:
I think it would be better for sirushti to update the documents after his PR is merged.

chemistry-sourabh replied:
@sirushtim please check the installation process

naved001 commented:
@chemistry-sourabh Thanks! I'll make these changes

chemistry-sourabh commented:
@naved001 are you waiting on something related to this PR?

naved001 commented:
@chemistry-sourabh @apoorvemohan updated the docs. have a look at it now.

chemistry-sourabh commented:
Still didn't read the installation part. Will ping you when done.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,283 @@
+For more information visit their [website](http://stgt.sourceforge.net/) and the [Quick Start Guide](https://fedoraproject.org/wiki/Scsi-target-utils_Quickstart_Guide)
+
+#### DHCP server
+You can setup `dnsmasq` for this. 
```
This is mandatory and the only one we support as of now.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,283 @@
+
+#### TFTP Server
+##### On ubuntu:
+* Install these packages:  `$ sudo apt-get install xinetd tftpd tftp`
```
Why do we need a explicit TFTP, dnsmasq is going to bring that in. Unless, we go for another dnsserver, this section should be optional.

naved001 replied:
Ah, makes sense. I have made this section optional. I am fine with removing it too.

chemistry-sourabh commented:
@ravisantoshgudimetla you also better hit the request changes button

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,283 @@
+verbose = False
+```
+
+### Einstein  
````
We need to explain what Einstein is, if not just a reference here.

naved001 replied:
I have put one line to tell what is Einstein and Picasso.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,283 @@
+
+Einstein is running!!
+
+### Picasso
```
Same applies for picasso as well.

ravisantoshgudimetla commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,111 @@
+Once the checklist above has been met, open a Pull Request from your
+personal fork to the main BMI repo.
+
+
```
Add a reference to git pull request(Things like forking and raise a PR from branch etc).

chemistry-sourabh replied:
you can add the link to the atlassian workflow tutorial.

naved001 replied:
I have linked the git workflow in Submitting code/Pull Requests section

naved001 commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,111 @@
+  - [ ] Ensure tests pass (coming soon!) after making your changes by running `py.test tests/unit ` from the top-level ims directory. Parallel testing can be used on multi-core systems by running `py.test tests/unit -n auto`
+  - [ ] Add unit tests in the corresponding file and create one if none are present.
+  - [ ] If practical, bug fixes should have an reproducing test to ensure that the bug does not come back.
+  - [ ] Run deployment tests if needed
```
> What do you mean by deployment tests ? Do you mean on the actual hardware ? If so they will run as part of the checks raised by the PR.

Yes, on actual hardware. But I don't understand what do you mean when you say "will run as part of the checks raised by the PR"

> So removing here and adding there would be better

Makes sense. Ill remove it from here and add it to a different doc.

naved001 commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,111 @@
+  - [ ] If practical, bug fixes should have an reproducing test to ensure that the bug does not come back.
+  - [ ] Run deployment tests if needed
+
+#### Get agreement
```
I'll change it to "Get agreeement from the BMI team" (or if you have any other suggestion). Point being, before a developer starts coding, the team should, at least, agree with the idea to save coding effort.

chemistry-sourabh replied:
What you suggested is fine.

naved001 commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,111 @@
+
+Reviewers are the final guardians for good code quality. If you see a piece of
+complex code, that is probably where you want to spend a lot of your review
+time. Remember the 80/20 rule (80% of the bugs come from 20% of the code).
```
what checks? :S

chemistry-sourabh replied:
The pep8 and test cases that will run in openshift or travis

naved001 replied:
okay, updated that in the "Prior to the pull request" section. Line 52 and 52

naved001 commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,111 @@
+should be added that provide adequate coverage.
+
+If fixing a bug, a regression test should accompany the bug fix to ensure that
+the bug does not return.
```
Actually, this is pretty general stuff, I think we should let this be here. And the more detailed testing stuff (like how to run, and the test  structures etc) would go into the test docs.

chemistry-sourabh replied:
In that case Add the link at the ending of the section saying refer for more details

naved001 replied:
added a link to `testing.md`

naved001 commented:
@ravisantoshgudimetla @chemistry-sourabh @apoorvemohan I addressed your comments for this PR. Have a look at it, and if you are happy just squash and merge.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,287 @@
+That's it. Installation is done!
+***
+## Running
+
```
What does running mean here?

naved001 replied:
running bmi. I changed the title and moved this section where it makes more sense.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,287 @@
+```
+
+**Driver section**
+* net_isolator is the name of the network isolator to load
````
We should define, what network isolator is or atleast point to HIL documentation.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,287 @@
+**iscsi section**
+* ip is the ip of iscsi server on the provisioning network
+* password is the sudo password for the VM (Will be removed)
+```
````
Add a notification here saying it will be deprecated in future.

chemistry-sourabh replied:
It is already deprecated if I remember right

naved001 replied:
It already says "will be removed". @chemistry-sourabh if it's already deprecated then I'll just remove this line.

chemistry-sourabh replied:
It is removed in TGT, but ceph still uses it for map and unmap which should be remove. So let it be.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,287 @@
+Einstein is the backend that does the operations. 
+
+First export the variable BMI_CONFIG
+
```
First, export the path of BMI config file to environment variable(BMI_CONFIG).

ravisantoshgudimetla commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,105 @@
+### Approval criteria
+
+Pull Requests require at least 2 successful reviews, also known as "+1"'s, in
+order to be merged, with at least 2 core developers being involved in some way:
```
Or approve in github.

naved001 replied:
not sure what you mean here. It's understood that we are talking about approving and merging PRs in github.

chemistry-sourabh replied:
@ravisantoshgudimetla likes for everything to be explicit. Why not 2 approves instead of successful reviews ?

naved001 replied:
@chemistry-sourabh ah, good catch! when this was written there was no "approval" feature in github then. I'll update it.

ravisantoshgudimetla replied:
> Or approve in github.

Thats what I meant earlier, I should have been explicit :).

chemistry-sourabh commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,288 @@
+```  
+
+**rpc section**
+* name_server ip and port is the RPC NameServer's IP and Port
````
is the ip and port Name Server binds to ( Similar to RPC server also)

chemistry-sourabh commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,105 @@
+### Code review
+
+Code reviews help increase code quality by finding:
+* mistakes that can oftentimes be overlooked by a single developer
```
often times

naved001 replied:
this is so meta lol.

chemistry-sourabh replied:
lol googled it, looks like oftentimes is an actual word. English is truly a weird language.

naved001 replied:
actually, oftentimes is a valid word

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,288 @@
+For more information, visit the [official website](http://docs.ceph.com/docs/master/start/)
+
+You can follow [this guide](http://palmerville.github.io/2016/04/30/single-node-ceph-install.html) 
+to setup a ceph client on a VM
```
It seems the link points to single node ceph installation. What we need is a client.(and it will vary from from one OS to other). So, we should be explicit in mentioning it.(the link points to ubuntu and can work for all debian flavours, I guess).

naved001 replied:
good catch. I update this section.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,288 @@
+systemctl list-unit-files | grep -i tgt
+
+iptables -I INPUT -p tcp -m tcp --dport 3260 -j ACCEPT
+service iptables save
```
Again, we should talk about iptables, only if OS is RHEL family.

ravisantoshgudimetla commented:
I guess, OS related things need to be explicitly mentioned like "On RHEL family or on ubuntu etc", so that there are no ambiguities while installation. I think if we mention that as of now we are supporting RHEL family makes much sense. I will make one more pass in coming week.

naved001 commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,288 @@
+
+For more information about dnsmasq, you can look at the [wiki](https://wiki.debian.org/HowTo/dnsmasq)
+
+#### TFTP Server (Optional)
```
cool, got rid of it.

naved001 commented:
okay. changed some more stuff around here. let me know if I missed out on something.

chemistry-sourabh commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,240 @@
+
+To install ceph client on ubuntu:
+```
+apt-get install ceph-common
````
Remove the ubuntu related commands. We said we officially support RHEL, so no need for this. Also iSCSI section doesnt have an ubuntu section so better remove it.

naved001 replied:
cool. will get rid of ubuntu stuff.

chemistry-sourabh commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,240 @@
+We support `dnsmasq` for this. 
+
+* On ubuntu:
+`$ sudo apt-get install dnsmasq`
```
Same like above

chemistry-sourabh commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,109 @@
+
+#### Testing
+
+* [Testing document](testing.md)
```
Remove this

chemistry-sourabh commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,109 @@
+
+The exception to this is documentation, where we are a little more lenient in
+the interest of lowering the barrier to having better docs. For docs changes,
+typically one +1 is sufficient.
```
We are not following this para so should it be there ?

naved001 replied:
I think that is because we are in the early stages. Once we are done with 0.5 issues, we'll be following these.

chemistry-sourabh commented on docs/Developer-guidlines.md:
```diff
@@ -0,0 +1,109 @@
+
+Whomever provides the enabling +1 is responsible for clicking the merge button
+on github. If you do not have commit access, then please add your +1 to the PR
+and ask one of the Core Developers to complete the merge.
```
Even this is not followed. We are asking whoever is the team lead before merging.

naved001 replied:
^^same as above. Right now we are a bit more paranoid which is why we are asking the team lead and others for opinions on most PRs.

chemistry-sourabh commented:
ok then

chemistry-sourabh commented:
Go have a Beer!!

sirushtim commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+
+This section describes each config section along with sample  
+
+**bmi section**  
```
This section can get easily outdated. Filed a bug for this: https://github.com/CCI-MOC/ims/issues/105

chemistry-sourabh replied:
Ya probably we can separate it out later when the issue is addressed

ravisantoshgudimetla replied:
Probably a link to sample file in git(which will be updated periodically is good enough).

naved001 replied:
there's already a link to a sample config file with every section explained in it (it's more like a template).

chemistry-sourabh commented:
@ravisantoshgudimetla your approve is left

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+
+cp scripts/tgtd.service /usr/lib/systemd/system/
+systemctl daemon-reload
+systemctl list-unit-files | grep -i tgt
```
We can tell that this command is for checking if tgtd has been made as a deamon

naved001 replied:
good idea!

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+# this section is for basic bmi settings
+[bmi]
+# uid is given so that images dont clash in ceph pool
+uid = 5
```
Do, we need to explain it twice. It seems you already did a good job explaining what it is and why it is needed, either move those to comments to lines above the parameters or remove the comment line here.

chemistry-sourabh replied:
or as you suggested the enhanced explanations can be moved to the sample config file

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,231 @@
+```
+
+**Driver section**
+* net_isolator is the name of the network isolator to use. it connects the
````
The main idea is to have multi-tenancy while provisioning(securely).

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+# This section is for hil related config
+[driver]
+fs = ims.einstein.fs.ceph.driver
+net_isolator = not used
```
<not used> may sound odd. Can we change this to NULL or something much more meaningful.

naved001 replied:
I have set this to <default> to indicate that it's using the default net isolator we have right now.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,231 @@
+**rpc section**
+* name_server is the ip and port Name Server binds to
+* rpc_server is the ip and port RPC Server binds to (The end which calls einstein)
+```
````
We should talk about pyro or atleast have a link to it.(https://pypi.python.org/pypi/Pyro4)

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,231 @@
+ipxe_path = /var/lib/tftpboot/
+```  
+**http section**
+* bind_ip and port is the ip and port picasso should bind to
````
I am thinking out loud here. Should this be http section or API server section?

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+### Einstein  
+
+Einstein is the backend that does the operations. 
+
```
I think, this should be somewhere in the starting, where we talked about Einstein first.

naved001 replied:
This is the only place where we talk about Einstein. In this installation document, we first talk about installing and configuring the Prerequisite softwares, then configuring the config file, and finally running the BMI services (Einstein and Picasso)

ravisantoshgudimetla commented on docs/BMI-Installation.md:
````diff
@@ -0,0 +1,231 @@
+path = /home/bmi/logs/
+debug = False
+verbose = False
+```
````
Do we have some default values for every parameter, if yes, we should talk about the default values as well.

naved001 replied:
1 is a path and 2 of those are booleans, and those are self explanatory? aren't they? moreover, I am not sure if you can have a default value here, it's just for reference anyway.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,231 @@
+Picasso is running!!
+
+### Initing DB
+
```
Bootstrapping DB, may be a better term, nit again, could be ignored.

naved001 commented:
@ravisantoshgudimetla I have made some changes, let me know if this looks good.

ravisantoshgudimetla commented on docs/BMI-Installation.md:
```diff
@@ -0,0 +1,132 @@
+## Installation
+
+This document describes how to setup BMI. We officially support RHEL and it's
+variants, but it may also work on ubuntu (Debian) like OSes. 
```
NIT: OSes doesn't make sense. Probably expanding it is clear.
